engines/nancy/sound.cpp

```diff
@@ -0,0 +1,335 @@
+	if (description.name == "NO SOUND") {
+		return;
+	}
+
+	if (_mixer->isSoundHandleActive(_channels[description.channelID].handle)) {
+		_mixer->stopHandle(_channels[description.channelID].handle);
+	}
+
+	delete _channels[description.channelID].stream;
+	_channels[description.channelID].stream = nullptr;
+
+	_channels[description.channelID].name = description.name;
+	_channels[description.channelID].numLoops = description.numLoops;
+	_channels[description.channelID].volume = description.volume;
+
+
+	Common::SeekableReadStream *file = SearchMan.createReadStreamForMember(description.name + ".his");
+	if (file) {
+		_channels[description.channelID].stream = makeHISStream(file, DisposeAfterUse::YES);
+	}
+}
+
+void SoundManager::playSound(uint16 channelID) {
+	if (channelID > 32 || _channels[channelID].stream == 0)
+		return;
+
+	_channels[channelID].stream->seek(0);
+
+	_mixer->playStream(	_channels[channelID].type,
+						&_channels[channelID].handle,
+						Audio::makeLoopingAudioStream(_channels[channelID].stream, _channels[channelID].numLoops),
+						channelID,
+						_channels[channelID].volume * 255 / 100,
+						0, DisposeAfterUse::NO);
+}
+
+void SoundManager::playSound(const SoundDescription &description) {
+	if (description.name != "NO SOUND") {
+		playSound(description.channelID);
+	}
+}
+
+void SoundManager::pauseSound(uint16 channelID, bool pause) {
+	if (channelID > 32)
+		return;
+
+	if (isSoundPlaying(channelID)) {
+		g_system->getMixer()->pauseHandle(_channels[channelID].handle, pause);
+	}
+}
+
+void SoundManager::pauseSound(const SoundDescription &description, bool pause) {
+	if (description.name != "NO SOUND") {
+		pauseSound(description.channelID, pause);
+	}
+}
+
+bool SoundManager::isSoundPlaying(uint16 channelID) {
+	if (channelID > 32)
+		return false;
+	
+	return _mixer->isSoundHandleActive(_channels[channelID].handle);
+}
+
+bool SoundManager::isSoundPlaying(const SoundDescription &description) {
+	if (description.name == "NO SOUND") {
+		return false;
+	} else {
+		return isSoundPlaying(description.channelID);
+	}
+}
+
+void SoundManager::stopSound(uint16 channelID) {
+	if (channelID > 32)
+		return;
+
+	if (isSoundPlaying(channelID)) {
+		_mixer->stopHandle(_channels[channelID].handle);
+	}
+	_channels[channelID].name = Common::String();
+	delete _channels[channelID].stream;
+	_channels[channelID].stream = nullptr;
+}
+
+void SoundManager::stopSound(const SoundDescription &description) {
+	if (description.name != "NO SOUND") {
+		stopSound(description.channelID);
+	}
+}
+
+// Returns whether the exception was skipped
+void SoundManager::stopAllSounds() {
+	for (uint i = 0; i < 32; ++i) {
+		stopSound(i);
+	}
+}
+
+void SoundManager::stopAndUnloadSpecificSounds() {
+	// TODO missing if
+
+	for (uint i = 0; i < 10; ++i) {
+		stopSound(i);
+	}
+}
+
+void SoundManager::initSoundChannels() {
+	// Original engine hardcoded these and so do we
+	_channels[7].type = Audio::Mixer::kSpeechSoundType;
```

Comment:
Added a table since loops wouldn't really work that well with how randomly the channel types are scattered. Also, I'm assuming other games will have different channel types so I'll probably end up having to make separate tables for each title.